Add imagePullSecret to the `default` helm create template

donbowman opened the pull request:
This relates to [#3529](https://github.com/helm/helm/issues/3529).
It adds image.pullSecret to the default generate blank chart
when the user does `helm create`.

Signed-off-by: Don Bowman <don@agilicus.com>

<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Make sure to read the Contributing Guide before submitting your PR: https://github.com/helm/helm/blob/master/CONTRIBUTING.md
2. If this PR closes another issue, add 'closes #<issue number>' somewhere in the PR summary. GitHub will automatically close that issue when this PR gets merged. Alternatively, adding 'refs #<issue number>' will not close the issue, but help provide the reviewer more context.-->

**What this PR does / why we need it**:

**Special notes for your reviewer**:

**If applicable**:
- [ ] this PR contains documentation
- [ ] this PR contains unit tests
- [ ] this PR has been tested for backwards compatibility

donbowman committed:
Add imagePullSecret to the `default` helm create template

This relates to [#3529](https://github.com/helm/helm/issues/3529).
It adds image.pullSecret to the default generate blank chart
when the user does `helm create`.

Signed-off-by: Don Bowman <don@agilicus.com>

unguiculus commented on pkg/chartutil/create.go:
```diff
@@ -62,6 +62,7 @@ image:
   repository: nginx
   tag: stable
   pullPolicy: IfNotPresent
+  # pullSecret: my_secret
```
Instead of commenting it out, add an empty default.

unguiculus commented on pkg/chartutil/create.go:
```diff
@@ -189,6 +190,10 @@ spec:
         app.kubernetes.io/name: {{ include "<CHARTNAME>.name" . }}
         app.kubernetes.io/instance: {{ .Release.Name }}
     spec:
+      {{- if .Values.image.pullSecret -}}
```
`imagePullSecrets` is a list. I'd suggest to respect that so it is possible to actually configure a list.

I would not nest it under `image` since it is not related to that but a field that's part of the pod spec. This will become obvious as soon as you have multiple containers in the pod.

Here's an example:
* https://github.com/helm/charts/blob/master/stable/mongodb-replicaset/templates/mongodb-statefulset.yaml#L43-L47
* https://github.com/helm/charts/blob/master/stable/mongodb-replicaset/values.yaml#L25

unguiculus replied:
And to be consistent, please use `with`.

donbowman committed:
Make imagePullSecrets and array, move to root level

Changes as per PR https://github.com/helm/helm/pull/5271

- make imagePullSecrets be an array
- move imagePullSecrets to 'root' level of Values

Signed-off-by: Don Bowman <db@donbowman.ca>

unguiculus commented on pkg/chartutil/create.go:
```diff
@@ -189,6 +190,10 @@ spec:
         app.kubernetes.io/name: {{ include "<CHARTNAME>.name" . }}
         app.kubernetes.io/instance: {{ .Release.Name }}
     spec:
+      {{- with .Values.imagePullSecrets }}
```
Match the existing indentation style.

```suggestion
    {{- with .Values.imagePullSecrets }}
```

unguiculus commented on pkg/chartutil/create.go:
```diff
@@ -189,6 +190,10 @@ spec:
+      {{- with .Values.imagePullSecrets }}
+      imagePullSecrets:
+        {{- toYaml . | nindent 8 }}
+      {{- end }}
```
Match the existing indentation style.

```suggestion
    {{- end }}
```

donbowman committed:
Change indent of {{ with }}

Signed-off-by: Don Bowman <db@donbowman.ca>

donbowman commented:
is anything more needed of me?

donbowman commented:
is anything more needed of me?

donbowman commented:
@unguiculus is there anything more needed of me on this?
